gcc compilations for ARM64 can not initialize extern "C" variables.

RobertHenry6bev opened the pull request:
Some versions of clang targeted to ARM64 do not allow extern "C"
variables to be initialized at the point of extern "C" declarations,
although this many not be caught until the assembler of linker is run,
at which time an illegal relocation type is discovered.

Per advice from Jan Vorlicek.

RobertHenry6bev committed:
Clang compilations for ARM64 can not initialize extern "C" variables.

Some versions of clang targeted to ARM64 do not allow extern "C"
variables to be initialized at the point of extern "C" declarations,
although this many not be caught until the assembler of linker is run,
at which time an illegal relocation type is discovered.

Per advice from Jan Vorlicek.

am11 commented:
What does the warning message look like? This breaks windows builds.

BTW, gcc legs are also warning about these. Search for `initialized and declared ‘extern’` on this page https://dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_apis/build/builds/1569433/logs/866 to find all instances.

janvorli commented:
@RobertHenry6bev I've thought you've said it was a problem with GCC and not clang.

janvorli commented:
This warning actually looks like a GCC bug. `extern "C"` should be usable with variable declaration to express that the variable has C style linkage (naming convention). The reason it has broken only Windows builds is that C and C++ mangling of variable names differ while on Unix they are the same.
I wonder if GCC could be quieted using this way instead:
```C++
extern "C" 
{
    void *JIT_WriteBarrier_Loc = 0;
}
```

am11 commented:
Actually both clang and gcc warn about it: https://godbolt.org/z/Wsfn7nT5j. Clang has this warning suppressable (`Wextern-initializer`), while gcc doesn't. However, going by C11 `§6.9.2 External object definitions ¶4`, it's fine.

janvorli commented:
@am11 but that's in C. In C++, clang doesn't warn.

I've just verified that the syntax I've proposed above makes it work without warnings with GCC too:
```c++
extern "C" 
{
    void *JIT_WriteBarrier_Loc = 0;
}
```

am11 commented:
@janvorli, true, clang doesn't warn in C++ context, but gcc does. Your way works for both: https://godbolt.org/z/4j6dr4aqn

am11 commented:
Would be nice to fix [other 12 instances](https://dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_apis/build/builds/1569433/logs/866) of this warning as part of this PR.

jkotas commented:
You can also avoid this warning by splitting the declaration and initialization:

```
extern "C" void *JIT_WriteBarrier_Loc2;

void *JIT_WriteBarrier_Loc2 = 0;
```

This pattern is used in number of places in coreclr. For example, here: https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/gcheaputilities.h#L27-L31

RobertHenry6bev committed:
Change 6 initialized extern "C" declarations to split declaration and initialization.

This seems to remove 2*6==12 warning errors
  initialized and declared ‘extern’
some of which were errors with modern gcc on ARM64.

janvorli reviewed (approved):
LGTM, thank you!

jkotas commented on src/coreclr/vm/threads.cpp:
```diff
@@ -1124,7 +1124,11 @@ extern "C" void *JIT_WriteBarrier_Loc;
 #ifdef TARGET_ARM64
 extern "C" void (*JIT_WriteBarrier_Table)();
+
+extern void *JIT_WriteBarrier_Loc;
```
I think this needs to be extern "C".

janvorli replied:
Right, I am sorry for missing that when reviewing the change.

jkotas commented:
The build is failing with:
```
2022-01-27T20:24:38.2826031Z AsmHelpers.asm.obj : error LNK2019: unresolved external symbol JIT_WriteBarrier_Table_Loc referenced in function LazyMachStateCaptureState
2022-01-27T20:24:38.2835441Z   Hint on symbols that are defined and could potentially match:
2022-01-27T20:24:38.2853622Z     "void * JIT_WriteBarrier_Table_Loc" (?JIT_WriteBarrier_Table_Loc@@3PEAXEA)
```

RobertHenry6bev committed:
Correctly declare extern "C" ...  JIT_WriteBarrier_Loc and companion.
